File: CSharpCodeBase/components/services/events.lua
local Class = require 'utils.hump.class'

local EventSystem = Class({init = function (self)
      self.handlers = {}
end})

function EventSystem:AddListener(event, listener)
  if self.handlers[event] == nil then
    self.handlers[event] = {}
  end
  local idx = Tables.Find(self.handlers[event], listener)
  if idx < 0 then
    self.handlers[event][#self.handlers[event] + 1] = listener
  end
end

function EventSystem:RemoveListener(event, listener)
  if self.handlers[event] == nil then
    self.handlers[event] = {}
  end
  local idx = Tables.Find(self.handlers[event], listener)
  table.remove(self.handlers[event], listener)
end

function EventSystem:Event(e, params)
  if self.handlers[e] ~= nil then
    local handlers = self.handlers[e]
    for k, listener in pairs(handlers) do
      if listener.OnEvent ~= nil then
        listener:OnEvent(e, params)
      else
        listener(e, params)
      end
    end
  end
end

return EventSystem

File: CSharpCodeBase/components/services/events.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class events {
 var EventSystem = Class({init = public void (self){
       this.handlers = {}
 }})
 public void EventSystem:AddListener(event, listener){
   if(this.handlers[event] == null  ){ 
     this.handlers[event] = {}
   }
   var idx = Tables.Find(this.handlers[event], listener);
   if(idx < 0  ){ 
     this.handlers[event][#this.handlers[event] + 1] = listener;
   }
 }
 public void EventSystem:RemoveListener(event, listener){
   if(this.handlers[event] == null  ){ 
     this.handlers[event] = {}
   }
   var idx = Tables.Find(this.handlers[event], listener);
   table.remove(this.handlers[event], listener);
 }
 public void EventSystem:Event(e, params){
   if(this.handlers[e] != null  ){ 
     var handlers = this.handlers[e];
     for(k, listener in pairs(handlers) ){
       if(listener.OnEvent != null  ){ 
         listener:OnEvent(e, params);
       }else{
         listener(e, params);
       }
     }
   }
 }
 ystem}}